### ISSUE TITLE
Mean Field Dev - system.py

### ISSUE THREAD

**Issue by gefux** [tempoCollaboration/OQuPy#56]
Hi @piperfw!
As discussed the system.py module needs to be restructured. See my suggestion below.

Considering `System`, `TimeDependentSystem`, `TimeDependentSystemWithField`, and possibly `SystemWithField`: for each the data stored, the input and or output types are different and I don't see any common functionality that couldn't be an external function, I believe the right design is the simplest with each being an independent class. The only thing they really share is a system dimension, hence the `BaseSystem` class. The methods that use these objects have to differentiate between them and then use `system.liouvillian()`, `system.liouvillian(time)` or `system.liouvillian(time_1, time_2, state, field)` accordingly.

```python
class BaseSystem(BaseAPIClass):
    def __init__(
            self,
            dimension: int,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        pass # ToDo

    @property
    def dimension(self) -> ndarray:
        pass # ToDo

class System(BaseSystem):
    def __init__(
            self,
            hamiltonian: ndarray,
            gammas: Optional[List[float]] = None,
            lindblad_operators: Optional[List[ndarray]] = None,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        pass # ToDo

    def liouvillian(self) -> ndarray:
        pass # ToDo

    @property
    def hamiltonian(self) -> ndarray:
        pass # ToDo

    @property
    def gammas(self) -> List[float]:
        pass # ToDo

    @property
    def lindblad_operators(self) -> List[ndarray]:
        pass # ToDo


class TimeDependentSystem(BaseSystem):
    def __init__(
            self,
            hamiltonian: Callable[[float], ndarray],
            gammas: \
                Optional[List[Callable[[float], float]]] = None,
            lindblad_operators: \
                Optional[List[Callable[[float], ndarray]]] = None,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        pass # ToDo

    def liouvillian(self, t: float) -> ndarray:
        pass # ToDo

    @property
    def hamiltonian(self) -> Callable[[float], ndarray]:
        pass # ToDo

    @property
    def gammas(self) -> List[Callable[[float], float]]:
        pass # ToDo

    @property
    def lindblad_operators(self) -> List[Callable[[float], ndarray]]:
        pass # ToDo

class TimeDependentSystemWithField(BaseSystem):
    def __init__(
            self,
            hamiltonian: Callable[[float, complex], ndarray],
            field_eom: Callable[[float, ndarray, complex], complex],
            gammas: \
                Optional[List[Callable[[float], float]]] = None,
            lindblad_operators: \
                Optional[List[Callable[[float], ndarray]]] = None,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        pass # ToDo

    def liouvillian(self,
            t: float = None,
            tf: float = None,
            state: ndarray = None,
            field: complex = None) -> ndarray:
        pass # ToDo

    @property
    def hamiltonian(self) -> Callable[[float, complex], ndarray]:
        pass # ToDo

    @property
    def gammas(self) -> List[Callable[[float], float]]:
        pass # ToDo

    @property
    def field_eom(self) -> Callable[[float, float, ndarray, complex], complex]:
        pass # ToDo

    @property
    def lindblad_operators(self) -> List[Callable[[float], ndarray]]:
        pass # ToDo


class SystemChain(BaseAPIClass):
    pass # Stays unchanged!
```

**Comment by piperfw** [comment]
Hi @gefux 

Thanks again for the structure outline. I've made the changes on [piperfw/TimeEvolvingMPO.](https://github.com/piperfw/TimeEvolvingMPO/tree/mean-field-dev) The code is running, but tests have not been updated for the new syntax.

Comments:
- I've made at attempt at writing a coherent group of functions for use by the classes in input checking, but I expect they may be simplified (note in keeping with `dynamics.py` I added Type hints)
- the [suggested random matrix generation](https://github.com/tempoCollaboration/TimeEvolvingMPO/pull/42#discussion_r772303915_) from the previous PR was implemented
- arguments to the liouvillians etc. are now not optional. Happily, this workaround is no longer needed
- A first pass has been made of all docstrings

**Comment by gefux** [comment]
By the way: I think this is the way to go considering the lines of the Zen of Python:

> If the implementation is hard to explain, it's a bad idea.
> If the implementation is easy to explain, it may be a good idea.


My explaination:

> System, TimeDependentSystem, TimeDependentSystemWithField, have different input, different output and different usages such that they are (despite their common name) mostly independent classes. The only thing they have in common is that they all have a Hilbert space dimension, which is why we have a BaseSystem class with that property.

**Comment by piperfw** [comment]
I've moved the helper functions to the end of the file and updated the testing.

I noted that the `.liouvillian` methods had no type checking. I added some to that in `SystemWithField` (but not the other Systems) - this is in keeping with the type checking done for other 'user-facing' functions (i.e. not underscore) but perhaps at odds with Python convention of avoiding type checking (see Duck Typing). Let me know what you think - feel free to add type checking to the other classes here or remove that in `SystemWithField` (will require parts of some tests).

**Comment by gefux** [comment]
Hi @piperfw, you are right that the type checking we are doing at several places is somewhat unpythonic. For most methods, I don't see realistic scenarios where a call would not fail for "unexpected" objects. My motivation behind this type checking is to give the user good hints about that, such that debugging (for the user) gets easier.
However, in the `.liouvillian` method, I haven't added any type checking so far because this method is usually not directly called by a user, but from other objects. So feel free to remove the type checking in `TimeDependentSystemWithField.liouvillian`. 
PS: I think we should, at some point, overhaul the whole type checking anyway: 1) we should throw propper exceptions instead of making assert statements. 2) we should try to make tools like `mypy` happy. But I'd say, for now, let's not worry too much.

**Comment by piperfw** [comment]
That's fine - I've left in the Type checking in `TimeDependentSystemWithField.liouvillian` for now as it shouldn't do any harm and can serve as a reminder for us to revisit the topic later on.

**Comment by gefux** [comment]
changes are merged into the mean-field-dev branch #62
